clang/lib/DPCT/Utility.cpp

```diff
@@ -208,47 +208,47 @@ size_t calculateExpansionLevel(const SourceLocation Loc) {
 }
 
 // Get textual representation of the Stmt.
-std::string getStmtSpelling(const Stmt *S) {
+std::string getStmtSpelling(const Stmt *S, SourceRange ParentRange) {
   std::string Str;
   if (!S)
     return Str;
   auto &SM = dpct::DpctGlobalInfo::getSourceManager();
   SourceLocation BeginLoc, EndLoc;
-  auto StmtRange = getStmtExpansionSourceRange(S);
-  BeginLoc = StmtRange.getBegin();
-  EndLoc = StmtRange.getEnd();
-
-  // getDefinitionRange generally get more precise location.
-  // However, in some case, we prefer to use the macro expansion location.
-  // e.g. #define BLOCKDIM 8 + 8
-  // when migrating BLOCKDIM in kernel call like foo<<<BLOCKDIM,1,0>>>()
-  // we prefer BLOCKDIM then "8 + 8"
-  if (SM.getFileOffset(EndLoc) - SM.getFileOffset(BeginLoc) != 0) {
+
+  // Get the definition range of the parent
+  ParentRange =
+      getDefinitionRange(ParentRange.getBegin(), ParentRange.getEnd());
+  ParentRange.setEnd(
+      ParentRange.getEnd().getLocWithOffset(Lexer::MeasureTokenLength(
+          ParentRange.getEnd(), SM,
+          dpct::DpctGlobalInfo::getContext().getLangOpts())));
+
+  int Length = 0;
+
+  auto ParentRangeSize = SM.getFileOffset(ParentRange.getEnd()) -
+                         SM.getFileOffset(ParentRange.getBegin());
+  if (ParentRangeSize <= 0) {
+    // if ParentRange is invalid, getDefinitionRange is the best we can have
     auto DRange = getDefinitionRange(S->getBeginLoc(), S->getEndLoc());
     BeginLoc = DRange.getBegin();
     EndLoc = DRange.getEnd();
+    Length = SM.getFileOffset(EndLoc) - SM.getFileOffset(BeginLoc) +
+             Lexer::MeasureTokenLength(
+                 EndLoc, SM, dpct::DpctGlobalInfo::getContext().getLangOpts());
+  } else {
+    // if ParentRange is valid, find the expansion location in the ParentRange
+    auto Range =
+        getRangeInRange(S, ParentRange.getBegin(), ParentRange.getEnd());
+    BeginLoc = Range.first;
+    EndLoc = Range.second;
+    Length = SM.getFileOffset(EndLoc) - SM.getFileOffset(BeginLoc);
   }
 
-  int Length =
-      SM.getFileOffset(EndLoc) - SM.getFileOffset(BeginLoc) +
-      Lexer::MeasureTokenLength(
-          EndLoc, SM, dpct::DpctGlobalInfo::getContext().getLangOpts());
+
   Str = std::string(SM.getCharacterData(BeginLoc), Length);
```

Comment:
Suggest adding assert(Length > 0) before Line248 to detect the unexpected overflow。